### ISSUE TITLE
False positives with `ArgumentParameterMismatch`

### ISSUE THREAD

**Issue by Stephan202** [google/error-prone#491]
I've tried the `ArgumentParameterMismatch` checker on our code base and (so far) found only false positives. The methods that trigger the warning all seem to boil down to one of the following two patterns:

* Recursive methods where arguments "change place". For example:
    ```java
    final class Dummy {
        // Normally one would name these parameters `a` and `b`, but then the  
        // bug isn't triggered, since `ArgumentParameterMismatch` only 
        // considers variable names of length 4 or greater.
        int gcd(final int first, final int second) {
            if (second == 0) { 
                return first;
            }

            return gcd(second, first % second);
            // ^ Triggers "Did you mean 'return gcd(first, first % second);'?"
        }
    }    
    ```  
* Methods on immutable classes which return derivative instances, where the given parameter is processed before being passed on to the class' pure constructor:
    ```java
    @Immutable
    final class DummyModel {
        private final String name;
        private final ImmutableList<String> values;

        DummyModel(final String name, final ImmutableList<String> values) {
            this.name = name;
            this.values = values;
        }

        DummyModel withValues(final Iterable<String> newValues) {
            final ImmutableList<String> copy = ImmutableList.copyOf(newValues);
            return new DummyModel(this.name, copy);
            // ^ Triggers "Did you mean 'return new DummyModel(this.name, values);'?"
            // When `copy` is inlined, the warning is not triggered.
        }
    }    
    ```  

The second pattern is quite prevalent in our code base. In any case, I wonder whether `ArgumentParameterMismatch` can be made smarter in order to detect these cases.

(I have some vague ideas, but they've not crystallized to the point where I can formulate concrete suggestions for improvement. I apologize if this issue is deemed non-actionable as it stands.)

**Comment by eaftan** [comment]
The ArgumentParameterMismatch checker is an experiment to see if we can reproduce the results from this ICSE paper: http://mp.binaervarianz.de/icse2016-names.pdf

It's not ready for prime time.  I would suggest leaving it disabled.
